utils/fixed_decimal/src/decimal.rs

```diff
@@ -697,6 +697,450 @@ impl FromStr for FixedDecimal {
+                    // the lower magnitude to match it
+                    decimal.lower_magnitude = target_lowest_magnitude;
+                } else if decimal.lower_magnitude < 0 {
+                    // If the target magnitude was positive,
+                    // we may have truncated digits off the right side
+                    // of the decimal point, reset the lower magnitude
+                    decimal.lower_magnitude = 0;
+                }
+            }
+        }
+        Ok(decimal)
+    }
+    /// Internal function for parsing directly from floats using ryū
+    fn new_from_f64_raw(float: f64) -> Result<Self, Error> {
+        if !float.is_finite() {
+            return Err(Error::Limit);
+        }
+        // note: this does not heap allocate
+        let mut buf = ryu::Buffer::new();
+        let formatted = buf.format_finite(float);
+        Self::from_str(formatted)
+    }
+
+    /// Internal function to round off `n` digits
+    /// from the right
+    ///
+    /// `self` must have at least `n` digits
+    ///
+    /// This may end up adding a digit to the left!
+    ///
+    /// This will not change the number of significant digits, it simply exists
+    /// to *round* them (and will typically reduce the size of `self.digits`)
+    ///
+    /// This function is responsible for fixing `digits`, `magnitude`, and `upper_magnitude`.
+    /// It will only modify upper_magnitude when it is not large enough to fit the rounded number.
+    /// The caller may fix up `lower_magnitude` by whatever scheme it desires
+    fn round_trailing_digits(&mut self, n: u16, mode: RoundingMode) -> Result<(), Error> {
+        debug_assert!(
+            self.digits.len() >= n as usize,
+            "Attempted to round off {} digits of number that has only {}",
+            n,
+            self.digits.len()
+        );
+        if n == 0 {
+            // no point attempting to round off any digits
+            return Ok(());
+        }
+
+        let cutoff = self.digits.len() - n as usize;
+
+        match mode {
+            RoundingMode::Unnecessary => {
+                for digit in &self.digits[cutoff..] {
+                    if *digit != 0 {
+                        return Err(Error::Limit);
+                    }
+                }
+                return Ok(());
+            }
+            RoundingMode::Truncate => {
+                self.digits.truncate(cutoff);
+                return Ok(());
+            }
+            // continue to rest of routine
+            RoundingMode::HalfExpand => (),
+        }
+
+        // Do we need to round our significant digits?
+        let round = self.digits[cutoff] >= 5;
+
+        self.digits.truncate(cutoff);
+
+        if round {
+            for digit in self.digits[..cutoff].iter_mut().rev() {
+                if *digit == 9 {
+                    // We need to round the next digit
+                    *digit = 0;
```

Comment:
Issue: An invariant of `FixedDecimal` is that `self.digits` is not allowed to contain trailing zeros, as documented in the struct.

This is because all logic involving adding, removing, or iterating over trailing zeros assumes that the number of trailing zeros is exactly equal to the difference between `self.lower_magnitude` and `self.magnitude - self.digits.len() + 1`.

Note also that truncating could "crack open" zeros hiding in the middle of `self.digits`; for example, "1.03" rounded to magnitude -1 results in "1.0", but you need to remove the trailing zero from the digits array.

In ICU, this is done with a helper function.

Unless you think that this restriction shouldn't be part of the FixedDecimal data model?

Reply:
Ah, this seems fine then, in this case I can always return an error

Reply:
Hm?  Your reply doesn't make sense for the issue I raised.

Reply:
Yeah sorry, I fixed a version of this problem elsewhere and replied in the wrong spot: fixed it here too now.